### PR TITLE
feat: add JSON serialization for subprocess IPC

### PR DESCRIPTION
**Pull request by Blackgaurd**
Adds `serialization.py` with `serialize_input()` and `deserialize_output()` for converting builder inputs/outputs to/from JSON. This is needed for the upcoming subprocess.Popen IPC switch since we can no longer use pickle-based multiprocessing.Queue.

Includes round-trip tests for datetime key serialization, nested dependency structures, and edge cases.

### CODE REVIEW

**Comment by Blackgaurd** [comment]
* **#170** <a href="https://app.graphite.com/github/pr/Blackgaurd/datastream-rs/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#171](https://github.com/Blackgaurd/datastream-rs/pull/171) <a href="https://app.graphite.com/github/pr/Blackgaurd/datastream-rs/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#172](https://github.com/Blackgaurd/datastream-rs/pull/172) <a href="https://app.graphite.com/github/pr/Blackgaurd/datastream-rs/172?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#169** <a href="https://app.graphite.com/github/pr/Blackgaurd/datastream-rs/169?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#168** <a href="https://app.graphite.com/github/pr/Blackgaurd/datastream-rs/168?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#167** <a href="https://app.graphite.com/github/pr/Blackgaurd/datastream-rs/167?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/Blackgaurd/datastream-rs/167?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: builders/server/runtime/serialization.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        "builder_path": str(builder_path),
+        "script_dir": str(script_dir),
+        "dependencies": serializable_deps,
+        "timestamp": timestamp.isoformat(),
+        "env_file": str(env_file) if env_file is not None else None,
+    }
+    return json.dumps(payload).encode("utf-8")
+
+
+def deserialize_output(json_bytes: bytes) -> tuple[str, Any]:
```
Comment (Blackgaurd) [inline builders/server/runtime/serialization.py]: wrap the output type of this result in a lightweight frozen dataclass

**Comment by Blackgaurd** [comment]
### Merge activity

* **Mar 15, 10:28 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/Blackgaurd/datastream-rs/167).
* **Mar 15, 10:28 PM UTC**: @Blackgaurd merged this pull request with [Graphite](https://app.graphite.com/github/pr/Blackgaurd/datastream-rs/167).
